### PR TITLE
Add UDP Request/Response for JSON formatted server status

### PR DESCRIPTION
**Pull request by Rob-NY**
Implement two UDP messages (request and response) to send JSON based server status message containing information not readily available through other means.  The JSON status includes connected client information including IP address as well as the current "real" recording state of the Jamulus server.   Further, a command line option is specified which both enables/disable this functionality as well as limits the IP addresses to which this JSON data can be sent.

The use cases to support this functionality rest in a server administrators ability to 1) Know the IP address of connected clients for the purpose of blocking (and subsequently re-admitting) clients; 2) To query the current recording state of the server.  Both of these will enable remote server control and monitoring in a Jamulus-ness way using existing protocols and message structures.

Live system using a version of Jamulus server with this PR is available at https://vjammers.com   (See #1847 and #1990)

JSON formatted server status message that includes the IP address of connected clients as well as the true, current recording state of the Jamulus server.   A command line option enables/disables this functionality and restricts the IP address that is permitted to request this status.

This change will require documentation.  I'm willing to write that documentation if PR is approved.

**Status of this Pull Request**

**What is missing until this pull request can be merged?**

Only tested on LINUX.   Do not have the ability to test on Mac or Windows.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want  (tested on Linux ONLY)
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [ ] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

**Comment by gilgongo** [comment]
Thanks for this - can you give it a more desriptive name (something like "JSON for recording status"?) and some more info on the use case for it?

**Comment by Rob-NY** [comment]
> Thanks for this - can you give it a more desriptive name (something like "JSON for recording status"?) and some more info on the use case for it?

I'm pretty new to using github with forked projects.  Please bear with me as I navigate the learning curve.  I've updated the commit name.  

There is a problem with the upstream merge for me that I think has to do with git self-resolving the CServer constructor changes incorrectly.  I'm working to fix this on my branch.

**Comment by gilgongo** [comment]
Perfect, thanks!

**Comment by softins** [comment]
Thanks for the work and the PR, @Rob-NY! While I appreciate that it solves a particular use case for you, I'm wondering if it would be better incorporated into the other JSON work being done in #1975, which uses TCP. Having two separate, different types of JSON interface (one UDP and one TCP) seems less than elegant. Disclaimer - I haven't yet had the time to study #1975.

**Comment by Rob-NY** [comment]
@softins  - Yes, I understand that concern and agree that there may need to be a choice made here.   While #1975 beat me to the PR, I surfaced my suggestion in June.  Further, my approach does **not** introduce a new messaging protocol (TCP), and it behaves in **exactly** the same way as all other Jamulus messages.   Further, it only "fills the gaps" for things that are not available today - specifically retrieving attributes of the connection information and the current recording state of the server.

**Comment by ann0see** [comment]
Hi, thanks for opening this PR. Definitely the code looks smaller and probably easier to maintain.

I think it’s speed vs features vs maintenance vs potential bugs. 
Can you cross review the other PR and @dtinth cross review yours? All my comments from the other PR are probably also valid here.

**Comment by ann0see** [comment]
@Rob-NY the styling Check failed. Could you please run clang-format?

File: src/main.cpp [hunk 1]
```diff
@@ -984,6 +996,11 @@ QString UsageArguments ( char** argv )
            "  -w, --welcomemessage  welcome message to display on connect\n"
            "                        (string or filename)\n"
            "  -z, --startminimized  start minimizied\n"
+           "      --serverpublicip  specify your public IP address when\n"
+           "                        running a slave and your own directory server\n"
+           "                        behind the same NAT\n"
+           "      --serverbindip    specify the IP address the server will bind to\n"
+           "      --notifyserver    specify server to receive UDP notifications IP:PORT\n"
```
Comment (pljones) [inline src/main.cpp:1003]: I'd suggest grabbing the short option `-N` and inserting this in the right order in the help text (alphabetical by short option, then any closely related options without short options).

File: src/server.cpp [hunk 2]
```diff
@@ -386,6 +389,25 @@ CServer::CServer ( const int          iNewMaxNumChan,
         WriteHTMLChannelList();
     }
 
+    // Notify Server Setup
+    if ( !strNotifyServerAddr.isEmpty() )
+    {
+
+        // No IPv6 here
```
Comment (pljones) [inline src/server.cpp:396]: Why?

File: src/server.cpp [hunk 3]
```diff
@@ -386,6 +389,25 @@ CServer::CServer ( const int          iNewMaxNumChan,
...
+
+        if ( NetworkUtil::ParseNetworkAddress ( strNotifyServerAddr, addrNotifyServer, false ) )
+        {
+            qInfo() << "-Server notifications sent to:" << addrNotifyServer.toString();
```
Comment (pljones) [inline src/server.cpp:400]: > `-Server`
Missing space.

File: src/server.cpp [hunk 4]
```diff
@@ -1702,6 +1753,60 @@ void CServer::SetWelcomeMessage ( const QString& strNWelcMess )
     strWelcomeMessage = strWelcomeMessage.left ( MAX_LEN_CHAT_TEXT );
 }
 
+/**
+ * Build a server status json string that can be used to either
+ * write to a disk-based status file, or sent via CL message.
+ */
+
+void CServer::BuildServerStatusJson ( QString& strServerStatus )
```
Comment (pljones) [inline src/server.cpp:1761]: What's the maximum size of this JSON object?

**Comment by ann0see** [comment]
Converting to draft for now.

**Comment by ann0see** [comment]
@Rob-NY if you want, you can add yourself as contributor here: https://github.com/jamulussoftware/jamulus/blob/89b3d0e9604aac7a8b2d241e351e82247b93457e/src/util.cpp#L493 and - if you want - you can add your project to the [RELATED-PROJECTS.md](http://related-projects.md/) file. We really appreciate your contribution, but it seems as if the JSON-RPC approach has more support. However - since you contributed valuable ideas, we would like to honor your work.

**Comment by Rob-NY** [comment]
That is very kind of you.  Thank you.

**Comment by ann0see** [comment]
@Rob-NY would you be ok in closing this in favor of JSON-RPC? If anything from this is needed in future, I think it would be good to have it on your repo (i.e. don't delete it). 

Nevertheless, thank you very much for your efforts and contributions for both API-approches ;-).

**Comment by Rob-NY** [comment]
Happy to.  I’m still maintaining this for my own use and need to make it available under the license, but it probably doesn’t need to sit as a PR.   I’m not a GitHub expert so let me know how to “close” or “cancel” this or if you can do it, go ahead.

**Comment by ann0see** [comment]
I'll close it for you. Thanks anyways (and it's still available on your repo so not lost).

**Pull request closed without merging**